apollo-federation/src/correctness/subgraph_constraint.rs

```diff
@@ -0,0 +1,155 @@
+// Path-specific constraints imposed by subgraph schemas.
+
+use std::sync::Arc;
+
+use apollo_compiler::collections::IndexMap;
+use apollo_compiler::collections::IndexSet;
+use apollo_compiler::executable::Field;
+
+use super::response_shape::NormalizedTypeCondition;
+use super::response_shape::PossibleDefinitions;
+use super::response_shape_compare::PathConstraint;
+use super::CheckFailure;
+use crate::error::FederationError;
+use crate::internal_error;
+use crate::link::federation_spec_definition::get_federation_spec_definition_from_subgraph;
+use crate::schema::position::CompositeTypeDefinitionPosition;
+use crate::schema::position::ObjectTypeDefinitionPosition;
+use crate::ValidFederationSchema;
+
+pub(crate) struct SubgraphConstraint<'a> {
+    /// Reference to the all subgraph schemas in the supergraph.
+    subgraphs_by_name: &'a IndexMap<Arc<str>, ValidFederationSchema>,
+
+    /// possible_subgraphs: The set of subgraphs that are possible under the current context.
+    possible_subgraphs: IndexSet<Arc<str>>,
+
+    /// subgraph_types: The set of object types that are possible under the current context.
+    /// - Note: The empty subgraph_types means all types are possible.
+    subgraph_types: IndexSet<ObjectTypeDefinitionPosition>,
+}
+
+/// Is the object type resolvable in the subgraph schema?
+fn is_resolvable(
+    ty_pos: &ObjectTypeDefinitionPosition,
+    schema: &ValidFederationSchema,
+) -> Result<bool, FederationError> {
+    let federation_spec_definition = get_federation_spec_definition_from_subgraph(schema)?;
+    let key_directive_definition = federation_spec_definition.key_directive_definition(schema)?;
+    let ty_def = ty_pos.get(schema.schema())?;
+    let mut resolvable_key_applications = Vec::new();
+    for directive in ty_def.directives.get_all(&key_directive_definition.name) {
+        let key_directive_application =
+            federation_spec_definition.key_directive_arguments(directive)?;
+        if !key_directive_application.resolvable {
+            continue;
+        }
+        resolvable_key_applications.push(key_directive_application);
+    }
+    Ok(!resolvable_key_applications.is_empty())
```

Comment:
I don't think you need a `Vec` here. We only need the first resolvable key application. `map().find_ok().map(Option::is_some)` should suffice.